### PR TITLE
feat: migrate to API rewrite

### LINKED ISSUE #224: Current Week PPAPI
Use PPAPI's week endpoint to get the current week.

- [ ] Remove current implementation (api\src\helpers\week.ts)
- [ ] Change endpoint (api\src\controllers\schedule.ts) to use [ppapi implementaiton](https://api.peterportal.org/REST-API/schedule/#scheduleweek)

### LINKED ISSUE #238: Investigate : updating course card prerequisite information
It seems that some of the prerequisite information is incorrect, perhaps this is due to the data for course prerequisites being outdated.  This issue is also present in AntAlmanac.

This task should include investigating the cause of this and deciding how we can improve frequency of updating prerequisite information.

image
^ example for CS 165 showing incorrect prerequisite
- CSE 45C OR I&C SCI 45C re only recommended prerequisites

Original issue submission for reference: #236

### LINKED ISSUE #301: Apostrophe decoding error with professor title
Apostrophe appears as `&#039;` as seen in the image below.

apostrophe

### LINKED ISSUE #343: Duplicate Prerequisites
Screenshot 2023-08-08 at 3 46 23 PM

Stats 67 requires "AP CALCULUS BC" whereas Math 3A requires "AP Calculus BC" resulting in what is essentially duplicate prerequisites which both need to be entered. Will need to investigate further to determine where the error is originating from, but my guess would be that WebSOC is not particularly consistent with how they write their Prereqs...

### LINKED ISSUE #344: Migrate to PPAPI rewrite
Migrate to PPAPI rewrite once its ready.

Should resolve the following data inconsistency issues:
- #343 
- #238 
- #301 

and should unblock #224

### LINKED ISSUE #350: Missing Bio Sci 48
When searching for classes to add Bio Sci 48 does not show up as an option.

### PR DESCRIPTION
**Pull request by ecxyzzy**
## Summary

Migrate PeterPortal Client to the API rewrite. [~~After [too long] in development, hopefully it will have been worth the wait.~~](https://knowyourmeme.com/memes/gaben#origin)

TODO before merging: remove the hardcoded API URLs and update the CI variables to point to the API rewrite.

## Test Plan

Overall: Check that the site still works like it does before, and doesn't crash. I tested most cases I could think of, but there are probably some edge and corner cases that I missed since I'm not that familiar with the codebase.

Specific issues:

- #238: Check that the prerequisites are correct.
- #301: It seems that Prof. Zhao no longer teaches at UCI (as of the last instructor scrape) so the case cannot be reproduced directly. However, UCI maintains a list of Chancellor's Professors [here](https://ap.uci.edu/titles-of-distinction/chancellors-professors/), and all of them seem to have the apostrophe rendered correctly.
- #343: Check that Stats 67 and Math 3A share "AP CALCULUS BC" as a prerequisite, and that one entry in Peter's Roadmap satisfies the prerequisite for both courses.
- #350: Check that Bio Sci 48 shows up in the courses tab.

## Known Issues and Next Steps

- ~~#352~~
#305
- #353

## Issues

Closes #224; closes #238; closes #301; closes #343; closes #344; closes #350.

### CODE REVIEW

File: site/src/types/types.ts
Comment (js0mmer) [inline site/src/types/types.ts:31]: Both CourseData and ProfessorData seem to be unused so feel free to remove them.

**Review by js0mmer — COMMENTED** [review]
I think these are all of the issues. Lmk if you have any questions or need help with anything. Thanks a ton!

File: site/src/pages/SearchPage/CourseHitItem.tsx
Comment (js0mmer) [reply site/src/pages/SearchPage/CourseHitItem.tsx:1]: Actually, only seems to be missing for some of them. COMPSCI classes seem to have the terms listed.

Not sure why the terms are missing from the AC ENG courses though, seems to be an API issue.
image
The course is still offered if you search it on WebSoc and was offered in 2020.

Note: the tooltip only shows 2020 terms because it was hardcoded for 2020 terms 💀. The fix will be part of #295

Comment (ecxyzzy) [reply site/src/pages/SearchPage/CourseHitItem.tsx:1]: Tracked in icssc/peterportal-api-next#113.

**Comment by ecxyzzy** [comment]
The term indicator issue is tracked on the API end, and I can write a solution for it before this gets merged if needed, but most of the substantial bugs and blockers should be resolved.

File: site/src/component/ReviewForm/ReviewForm.tsx
Comment (js0mmer) [inline site/src/component/ReviewForm/ReviewForm.tsx:1]: The review form seems to have broken itself again. Professor id is missing on form submission.
image

This is like the 3rd time it's broken in someone's branch, idk how. I think I've fixed it in my [vite PR](https://github.com/icssc/peterportal-client/pull/375/files#diff-da1836c6734a6c358bd863945cfca71254c5d9fdae560ac28862cc319b0a6e1cR164-R195) though. I just replaced the document.getElementsByName calls and id attribute with the actual sane approach of using the value attribute.

How do you want to do this? My PR is ready for merging but I don't want to leave you with too many merge conflicts to handle if I merge first.

**Comment by js0mmer** [comment]
> The term indicator issue is tracked on the API end, and I can write a solution for it before this gets merged if needed, but most of the substantial bugs and blockers should be resolved. Thanks for the help and sorry about the delay.

All good, it can wait since we still have an outstanding issue on our end to make the term indicator relevant.

**Review by js0mmer — CHANGES_REQUESTED** [review]
Thanks, search page is much faster now and mostly everything looks to be resolved.

I did notice that AP classes in the prerequisite tree have nonexistent links to course pages but that was preexisting problem so I opened a separate issue for it (#385)

File: site/src/component/PrereqTree/PrereqTree.tsx
Comment (ecxyzzy) [reply site/src/component/PrereqTree/PrereqTree.tsx:187]: Oops, must have glossed over it while resolving merge conflicts :facepalm:

File: site/src/component/ReviewForm/ReviewForm.tsx
Comment (ecxyzzy) [reply site/src/component/ReviewForm/ReviewForm.tsx:1]: Does this currently happen in prod? If so I think it would be best to isolate this into a separate PR and merge it into main, at which point I can merge it into my branch.

Otherwise we can probably wait until this get merged; I don't think it's a good idea for me to handle more merge conflicts considering I literally undid one of your member's changes :skull:

File: api/src/controllers/courses.ts
Comment (ecxyzzy) [reply api/src/controllers/courses.ts:42]: Yeah it does feel a bit slower, I will have to look into how best to resolve this.

The reason I used a batch of REST queries rather than a single composite GraphQL query (like before) is because currently the entire GraphQL query fails if a course does not exist. This can be resolved on the API end by making the `course` query return a nullable course; if that ends up being better performance wise, I'll update prod to reflect this behavior.

File: site/src/component/ReviewForm/ReviewForm.tsx
Comment (js0mmer) [reply site/src/component/ReviewForm/ReviewForm.tsx:1]: It does not happen in prod. You wanna just copy the changes I did in my PR to here (adjusted to the different property names on the course/professor objects)?

File: api/src/controllers/courses.ts
Comment (js0mmer) [reply api/src/controllers/courses.ts:42]: Got it. To clarify, I assume this is occurring on the course page when a course that no longer exists is present in the prerequisite tree. (e.g. CSE45C as a prerequisite for ICS46). Or are there other scenarios where we request information for a nonexistent course (I assume our search is fine and websoc-fuzzy-search only returns existing courses)?

Comment (ecxyzzy) [reply api/src/controllers/courses.ts:42]: Yeah pretty much; fuzzy search should only return courses that we have data for.

**Comment by ecxyzzy** [comment]
The latest staging build now points to an API staging instance which has the GraphQL change. It still seems quite slow to me which makes me think this is a symptom of a larger issue. Could you confirm whether this is the case?

**Comment by js0mmer** [comment]
Yeah, still seems slow.

**Comment by js0mmer** [comment]
Also, the review form is still broken. Seems to just be that the GraphQL response for professors is not mapped properly.
image
ucinetid should be the keys

**Comment by js0mmer** [comment]
Some really rough performance measurements from throwing print statements in a few places. In general, API rewrite does seem to be a lot slower at responding for course queries, instructors queries seem fine from what I can tell (this is with the batch REST requests for courses and single GraphQL query for instructors). Some improvement could probably be done on our end as well so I might draft up a PR for that.

API time is time measured from right before the request(s) are sent by the express server to after they're all received. Total was measured around the frontend's request to our express server (and maybe some other stuff in there, again really rough measurement).

# api rewrite
## search page

| query  | courses total search time (ms) | courses API time (ms) | professors total search time (ms) | professors api time (ms) |
| ------- | --- | --- | --- | --- |
| Empty   | 1011 | 475 | 1997 | 5 |
| compsci | 769 | 320 | | |
| mike    | | | 364 | 1.29 |
| compsci 121 | 668 | 241 |

## course page

| course | total time (ms) | API time (aggregated) (ms) |
| --- | --- | --- |
| compsci 121 | 1252 | 232 + 242 |

## professor page

| professor | total time (ms) | API time (aggregated) (ms) |
| --- | --- | --- |
| mikes | 1655 | 3 + 826 |

# original
## search page

| query  | courses total search time (ms) | courses API time (ms) | professors total search time (ms) | professors api time (ms) |
| ------- | --- | --- | --- | --- |
| Empty   | 540 | 3 | 547 | 3 |
| compsci | 332 | 3 | | |
| mike    |  | | 211 | 2.5 |
| compsci 121 | 233 | 2.5 | |

## course page

| course | total time (ms) | API time (aggregated) (ms) |
| --- | --- | --- |
| compsci 121 | 664ms | 3ms |

## professor page

| professor | total time (ms) | API time (aggregated) (ms) |
| --- | --- | --- |
| mikes | 1031 | 2.6 |

**Comment by ecxyzzy** [comment]
Thanks for running the numbers. They're definitely worse than I thought; while I suspect it is mostly cold start and/or DB overhead, it's still a pretty hefty penalty even if it doesn't happen that often. I'll look into this further after Wednesday since I should be done with everything except finals by then.

**Comment by js0mmer** [comment]
Ran a couple more tests on just the search page with GraphQL being used for the course queries. My conclusion is that we should definitely avoid any batch requests. I'm not too sure what went wrong on the course page though, I wouldn't think 3 requests would be an issue, might need to test again with nullable course query.

| query  | courses total search time (ms) | courses API time (ms) |
| ------- | --- | --- |
| Empty   | 855 | 1.8 |
| compsci | 962 | 2 |

**Review by js0mmer — COMMENTED** [review]
Something's up with the prisma set up on the staging instances. Following endpoints aren't working: 
* https://staging-117.api-next.peterportal.org/v1/rest/week
* https://staging-117.api-next.peterportal.org/v1/rest/websoc/terms
* https://staging-117.api-next.peterportal.org/v1/rest/grades/raw?department=COMPSCI&courseNumber=122B

Overall, does seem significantly faster. There is still the persisting slowdown with having to make the extra requests for prerequisite list, prerequisite for, and instructor history on course pages though, especially for those with a lot of prerequisites/prerequisite fors, such as Math 2B.

Below is for Math 2B (tested locally). Requests are for instructor history, prerequisite list, and prerequisite for respectively.
image
Is there a reason why you didn't include the nested GraphQL queries for the api rewrite? I feel like it's been causing a bit of a headache.

I can think of an improvement for this implementation though.
We are requesting much more information than needed for instructor history and prerequisites.
e.g. for each instructor in instructor history
```json
{
    "_0": {
        "name": "Natalia Komarova",
        "shortenedName": "KOMAROVA, N.",
        "ucinetid": "komarova",
        "title": "Professor",
        "department": "Mathematics",
        "schools": [
            "School of Biological Sciences",
            "School of Physical Sciences"
        ],
        "relatedDepartments": [
            "BIO SCI",
            "DEV BIO",
            "ECO EVO",
            "MOL BIO",
            "NEURBIO",
            "CHEM",
            "EARTHSS",
            "MATH",
            "PHY SCI",
            "PHYSICS"
        ],
        "courseHistory": {
            "MATH 2B": [
...
```
In the prior implementation with the nested queries, we only selected ucinetid, shortened name, and full name for each instructor. I assume doing the same here could result in significant improvements. Same goes for prerequisite list/for, we only selected id, department, number, and title. Whether or not this will be as fast as the single nested query though, I'm not sure.

**Comment by ecxyzzy** [comment]
> Something's up with the prisma set up on the staging instances. Following endpoints aren't working:

Thanks for pointing this out, I think I forgot to upgrade the query engine version to the one that the Node 20 lambda runtime supports. Pretty sure things should be working now though.

> Is there a reason why you didn't include the nested GraphQL queries for the api rewrite? I feel like it's been causing a bit of a headache.

Yeah now that I'm actually integrating the new API with PPC it seems like there was a reason that things were done a certain way. (Go figure.)

I think a way forward that we could consider is to have a separate sub-endpoint that returns nested entries for courses/instructors. I have a rough idea of how to implement this but will still need to work through it, and I'm unsure as to what the performance will look like compared to fetching everything, but it stands to reason that we could reduce the overhead of making multiple HTTP requests.

**Comment by ecxyzzy** [comment]
I think the performance issues should be resolved for good this time but can't tell for sure due to regional latency. Could you check on your end?

**Review by js0mmer — APPROVED** [review]
Performance looks good! Let me know when it's ready for merging and I'll update the actions variables. Thanks.

File: site/src/component/ReviewForm/ReviewForm.tsx
Comment (js0mmer) [inline site/src/component/ReviewForm/ReviewForm.tsx:144]:
```suggestion
```

**Pull request merged**
